File: apps/RelayiumKit/Sources/RelayiumAppKit/DeviceInbox/InboxMessageStore.swift
import Foundation
import Darwin
@preconcurrency import RelayiumKit

/// Where a received MESSAGE is kept, and why it is not a file.
///
/// ## The whole reason this type exists
///
/// A Device Inbox delivery is files or it is a message (`InboxManifestKind`),
/// and the two have nothing in common past the ciphertext. A message has no
/// name, no extension and no place in a folder somebody chose for their
/// downloads. A receiver that wrote one there as `message.txt` would deliver a
/// different thing from the one that was sent — and it would do it into a
/// directory the user shares, syncs or backs up on the strength of a decision
/// they made about FILES.
///
/// So a message is committed HERE: a per-account directory in Application
/// Support at 0700, one 0600 record per delivery. That has four consequences
/// worth naming, because each is a requirement rather than a side effect.
///
///  1. **It never touches the receive folder.** Nothing in this file resolves a
///     bookmark, opens a security scope, or takes a path from the user's grant.
///     A message therefore lands on a Mac whose receive folder is missing,
///     revoked, unwritable or simply never chosen — which is the point:
///     `InboxReceiver` classifies the delivery BEFORE it consults the folder,
///     and only a file delivery can be blocked by one.
///  2. **It is not purgeable.** Application Support, not `tmp` and not
///     `caches`: this is the user's own content and the system may not decide to
///     reclaim it. `IOSSurfaceGuardTests` already refuses the other two here.
///  3. **It is account-scoped**, for the same reason the journal and the key
///     history are. Two accounts on one Mac, and a message one of them received
///     must not be readable by, or survive into, the other's session.
///  4. **It is keyed by TASK ID**, which is what makes a replayed commit
///     idempotent. A crash between writing the record and journalling it leaves
///     the retry writing the SAME record at the SAME name, so a duplicate
///     delivery is impossible by construction rather than by a check that could
///     be forgotten.
///
/// ## What never leaves it
///
/// The message body. It is not logged, not put in a notification, not sent to
/// central, not written into the journal, and not carried on an `InboxReceipt`.
/// Everything downstream of a commit carries a byte COUNT and an id; a caller
/// that wants the text asks this store for it, which is a deliberate, auditable
/// call and not something a log line can do by accident.

/// One received message.
public struct InboxMessage: Equatable, Sendable, Identifiable {
    /// The delivery's task id. The de-duplication identity, and the file name.
    public let id: String
    /// When this Mac committed it.
    public let receivedAt: Date
    /// The message itself. Only ever read from disk by an explicit call.
    public let text: String

    public init(id: String, receivedAt: Date, text: String) {
        self.id = id
        self.receivedAt = receivedAt
        self.text = text
    }

    /// The length the manifest declared, and the only measure of a message that
    /// anything outside this file is allowed to render.
    public var byteCount: Int { text.utf8.count }
}

public enum InboxMessageStoreError: Error, Equatable, Sendable {
    /// A task id that could not safely become a file name.
    case invalidID
    /// Not one nonempty UTF-8 message within the protocol's bounds. Refused
    /// rather than truncated or repaired: a message this store cannot hold
    /// exactly is one it must not claim to have received.
    case invalidMessage
    /// The stored record is unreadable, or a version this build does not
    /// understand. Refused rather than parsed optimistically.
    case unreadable
    /// A filesystem call failed. Carries `errno` and nothing else — a path in an
    /// error is one step from a path in a log.
    case system(Int32)
}

/// The messages one account has received, in a 0700 directory.
public struct InboxMessageStore: Sendable {
    public let directory: URL

    public init(directory: URL) { self.directory = directory }

    /// **There is no retention, and no time-based deletion anywhere in this
    /// type.**
    ///
    /// This store briefly had both: a one-year cutoff and a `prune(now:)` the
    /// receive loop called on every idle pass. Both are gone, and their absence
    /// is a requirement rather than an omission waiting to be filled in.
    ///
    /// A journal entry is bookkeeping ABOUT a delivery and may expire once it
    /// can no longer prevent a duplicate or a data loss — `InboxJournalStore`
    /// still prunes for exactly that reason. A message IS the delivery. It is
    /// the user's own content, it exists in no other copy on this Mac, and no
    /// interval is long enough to make deleting it on a schedule they never
    /// chose anything other than silent data loss. A person who has not opened
    /// the app in fourteen months has not consented to losing what was sent to
    /// them; they have been away.
    ///
    /// `remove(_:)` below is the only deletion this type offers, it takes one
    /// id, and nothing in the receive loop calls it. Deletion is a thing a user
    /// does, so it needs a user.

    /// The stored-record format. Bumped only for a change this build could not
    /// read; an unrecognised version is refused, never guessed at.
    static let version = 1

    private struct Record: Codable {
        var version: Int
        var id: String
        var receivedAt: Int64
        var text: String
    }

    func path(_ id: String) throws -> URL {
        guard let checked = try? StoredObjectID.checked(id), checked.count <= 64 else {
            throw InboxMessageStoreError.invalidID
        }
        return directory.appendingPathComponent(checked + ".json")
    }

    /// Whether `text` is a message this store will accept.
    ///
    /// The same bounds the manifest declares, re-applied to the BYTES that
    /// actually arrived: the manifest is sender-controlled and says only what
    /// the sender intended to send. Nonempty because an empty message is not a
    /// message, and bounded at 64 KiB because that is the protocol's ceiling and
    /// this store holds a message in memory to hand it back.
    public static func isAcceptable(_ text: String) -> Bool {
        let bytes = text.utf8.count
        return bytes >= InboxManifest.minTextBytes && bytes <= InboxManifest.maxTextBytes
    }

    /// Commit one message, durably, and return only once it is on stable
    /// storage.
    ///
    /// Idempotent by construction: the record is named by task id, so a replayed
    /// commit rewrites the same record rather than adding a second one. That is
    /// the message half of the journal's no-duplicate contract — the caller
    /// records the commit in the journal AFTER this returns, and a crash in
    /// between costs a rewrite of identical bytes and nothing else.
    public func commit(id: String, text: String, receivedAt: Date) throws {
        guard Self.isAcceptable(text) else { throw InboxMessageStoreError.invalidMessage }
        let url = try path(id)
        let record = Record(version: Self.version, id: url.deletingPathExtension().lastPathComponent,
                            receivedAt: Int64(receivedAt.timeIntervalSince1970), text: text)
        guard let data = try? JSONEncoder().encode(record) else {
            throw InboxMessageStoreError.invalidMessage
        }
        do {
            try InboxJournalStore.ensureDirectory(directory)
            try InboxJournalStore.writeSecretFile(at: url, data: data, in: directory)
        } catch let error as InboxJournalError {
            // Re-typed rather than rethrown: the two stores fail for the same
            // POSIX reasons and a caller classifying this one must not have to
            // know it borrows the other's durable-write primitive.
            switch error {
            case .invalidTaskID: throw InboxMessageStoreError.invalidID
            case .unreadable: throw InboxMessageStoreError.unreadable
            case .system(let code): throw InboxMessageStoreError.system(code)
            }
        }
    }

    /// One message, or nil when this account has never received that delivery.
    public func load(_ id: String) throws -> InboxMessage? {
        let url = try path(id)
        guard let data = FileManager.default.contents(atPath: url.path) else {
            // Distinguish "not there" from "there and unreadable": only the
            // first is a delivery that never happened.
            var st = stat()
            if lstat(url.path, &st) == 0 { throw InboxMessageStoreError.unreadable }
            return nil
        }
        guard let record = try? JSONDecoder().decode(Record.self, from: data),
              record.version == Self.version, Self.isAcceptable(record.text) else {
            throw InboxMessageStoreError.unreadable
        }
        return InboxMessage(id: record.id,
                            receivedAt: Date(timeIntervalSince1970: TimeInterval(record.receivedAt)),
                            text: record.text)
    }

    /// Every message this account holds, newest first — ALL of them, however old.
    ///
    /// Unreadable records are SKIPPED rather than thrown: one corrupt file must
    /// not make the rest of a person's messages unreachable. Nothing else is
    /// skipped; there is no cutoff and no limit, because this is the call the
    /// surface renders and a message missing from it is a message the user has
    /// no way to reach.
    ///
    /// The tie-break on id is not decoration. `receivedAt` is stored to the
    /// second, so two deliveries worked in the same second compare equal — and
    /// `sorted(by:)` is not stable, so without it the newest-first order of a
    /// pair that landed together would differ between two reads of the same
    /// unchanged directory. The id ordering is arbitrary but FIXED, which is
    /// what a list that must not reshuffle under the reader needs.
    public func all() -> [InboxMessage] {
        guard let names = try? FileManager.default.contentsOfDirectory(atPath: directory.path) else {
            return []
        }
        return names
            .filter { $0.hasSuffix(".json") }
            .compactMap { try? load(String($0.dropLast(".json".count))) }
            .sorted {
                $0.receivedAt == $1.receivedAt ? $0.id > $1.id : $0.receivedAt > $1.receivedAt
            }
    }

    /// Delete ONE message, named by the caller.
    ///
    /// The only deletion this type has, and it is deliberately the shape that
    /// cannot become a policy: it takes a single id, so there is nothing here to
    /// hand a cutoff, a count or a predicate to. Nothing in the receive loop
    /// calls it — see the note above `version`.
    public func remove(_ id: String) throws {
        let url = try path(id)
        if unlink(url.path) != 0, errno != ENOENT {
            throw InboxMessageStoreError.system(errno)
        }
        InboxJournalStore.fsyncDirectory(directory)
    }
}

File: apps/android/app/src/main/kotlin/com/relayium/android/inbox/InboxMessageStore.kt
package com.relayium.android.inbox

import com.relayium.android.cloud.DurableFiles
import com.relayium.android.cloud.SecretBox
import com.relayium.android.cloud.SecretBoxException
import com.relayium.protocol.Json
import com.relayium.protocol.inbox.InboxManifest
import java.io.File
import java.io.IOException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

/**
 * Received messages, as messages.
 *
 * ## Why this is not a `.txt` file in the received directory
 *
 * `inbox.text.v1` is a promise to the SENDER that this receiver presents a text
 * delivery as text. A build that wrote messages into the receive folder as files
 * and announced the token would be making that promise falsely — and the whole
 * value of the token is that its absence is truthful. So a message is committed
 * here, keyed by task id, and read back by a surface that shows it.
 *
 * The practical consequence is the one the protocol intends: a message does not
 * depend on the receive directory at all. A container that is missing, full or
 * unwritable has nothing to do with whether a message can land, which is why the
 * receiver classifies kind BEFORE it consults the container.
 *
 * ## Retention
 *
 * A journal is bookkeeping ABOUT a delivery and expires once it can no longer
 * prevent a duplicate. A message IS the delivery — the user's own content — so
 * nothing here expires on a timer. It is removed when the user deletes it, or
 * when the account's data is destroyed.
 */
data class InboxMessage(
    val taskId: String,
    val senderDeviceId: String,
    val text: String,
    val receivedAt: Long,
) {
    /** The BODY never reaches a diagnostic: it is the user's own content. */
    override fun toString(): String =
        "InboxMessage(task=$taskId, bytes=${text.toByteArray(Charsets.UTF_8).size})"
}

enum class InboxMessageReason { INVALID_TASK_ID, UNREADABLE, STORAGE, MALFORMED_TEXT }

class InboxMessageException(val reason: InboxMessageReason, cause: Throwable? = null) :
    RuntimeException("relayium inbox message: $reason", cause)

/** One account's received messages, sealed at rest. */
class InboxMessageStore(
    val directory: File,
    private val account: InboxAccountId,
    private val secrets: SecretBox,
    private val files: DurableFiles = DurableFiles.Platform,
    private val io: CoroutineDispatcher = Dispatchers.IO,
) {

    private val lock = Mutex()

    /**
     * Commit one message, replacing any record under the same task id.
     *
     * Replacement is what makes this idempotent across a re-delivery: the store
     * is keyed by task id, so a crash between committing here and journalling it
     * costs a re-download that rewrites the SAME record at the SAME name. The
     * window produces a duplicate of nothing.
     */
    suspend fun commit(
        taskId: String,
        senderDeviceId: String,
        text: String,
        receivedAtSeconds: Long,
    ): InboxMessage = lock.withLock {
        if (!isAcceptable(text)) {
            throw InboxMessageException(InboxMessageReason.MALFORMED_TEXT)
        }
        val message = InboxMessage(taskId, senderDeviceId, text, receivedAtSeconds)
        write(message)
        message
    }

    suspend fun read(taskId: String): InboxMessage? = lock.withLock { readLocked(taskId) }

    /** Every readable message, newest first. An unreadable record is skipped
     *  rather than failing the list: one bad file must not hide the rest. */
    suspend fun all(): List<InboxMessage> = lock.withLock {
        withContext(io) { directory.list().orEmpty() }
            .filter { it.endsWith(SUFFIX) }
            .mapNotNull { name ->
                try {
                    readLocked(name.removeSuffix(SUFFIX))
                } catch (_: InboxMessageException) {
                    null
                }
            }
            .sortedByDescending { it.receivedAt }
    }

    /** Remove one message. The user's decision, never a timer's. */
    suspend fun delete(taskId: String): Boolean = lock.withLock {
        withContext(io) {
            val file = file(taskId)
            val existed = file.exists()
            if (existed && !file.delete()) {
                throw InboxMessageException(InboxMessageReason.STORAGE)
            }
            if (existed && directory.isDirectory) {
                try {
                    files.syncDirectory(directory)
                } catch (e: IOException) {
                    throw InboxMessageException(InboxMessageReason.STORAGE, e)
                }
            }
            existed
        }
    }

    // ── storage ─────────────────────────────────────────────────────────────

    private fun file(taskId: String): File = try {
        File(directory, InboxId.checked(taskId, "taskId") + SUFFIX)
    } catch (e: InboxWireException) {
        throw InboxMessageException(InboxMessageReason.INVALID_TASK_ID, e)
    }

    /** Binds the account AND the task, so a record moved between either fails to
     *  open rather than being presented under the wrong conversation. */
    private fun label(taskId: String) = "$LABEL_PREFIX/${account.value}/$taskId"

    private suspend fun readLocked(taskId: String): InboxMessage? = withContext(io) {
        val file = file(taskId)
        if (!file.exists()) return@withContext null
        val sealed = try {
            if (file.length() > MAX_RECORD_BYTES) {
                throw InboxMessageException(InboxMessageReason.UNREADABLE)
            }
            file.readBytes()
        } catch (e: IOException) {
            throw InboxMessageException(InboxMessageReason.STORAGE, e)
        }
        val plaintext = try {
            secrets.open(label(taskId), sealed)
        } catch (e: SecretBoxException) {
            throw InboxMessageException(InboxMessageReason.UNREADABLE, e)
        }
        try {
            decode(String(plaintext, Charsets.UTF_8), taskId)
        } finally {
            plaintext.fill(0)
        }
    }

    private suspend fun write(message: InboxMessage) = withContext(io) {
        val plaintext = encode(message).toByteArray(Charsets.UTF_8)
        val sealed = try {
            secrets.seal(label(message.taskId), plaintext)
        } catch (e: SecretBoxException) {
            throw InboxMessageException(InboxMessageReason.STORAGE, e)
        } finally {
            plaintext.fill(0)
        }
        try {
            files.createDirectories(directory)
            files.writeAtomically(file(message.taskId), sealed)
        } catch (e: IOException) {
            throw InboxMessageException(InboxMessageReason.STORAGE, e)
        }
    }

    private fun encode(message: InboxMessage): String = Json.stringify(
        Json.obj(
            "version" to Json.of(VERSION),
            "taskId" to Json.of(message.taskId),
            "senderDeviceId" to Json.of(message.senderDeviceId),
            "text" to Json.of(message.text),
            "receivedAt" to Json.of(message.receivedAt),
        ),
    )

    private fun decode(text: String, taskId: String): InboxMessage {
        val root = Json.parseOrNull(text) as? Json.Obj
            ?: throw InboxMessageException(InboxMessageReason.UNREADABLE)
        fun str(key: String) = (root[key] as? Json.Str)?.value
            ?: throw InboxMessageException(InboxMessageReason.UNREADABLE)
        val version = (root["version"] as? Json.Num)?.value
        if (version == null || version != VERSION.toDouble()) {
            throw InboxMessageException(InboxMessageReason.UNREADABLE)
        }
        // The file name is the task id; a record naming another task was moved,
        // and presenting it here would attribute one conversation's message to
        // another.
        if (str("taskId") != taskId) throw InboxMessageException(InboxMessageReason.UNREADABLE)
        val body = str("text")
        if (!isAcceptable(body)) throw InboxMessageException(InboxMessageReason.UNREADABLE)
        val receivedAt = (root["receivedAt"] as? Json.Num)?.value
            ?: throw InboxMessageException(InboxMessageReason.UNREADABLE)
        if (!receivedAt.isFinite() || receivedAt != Math.floor(receivedAt) || receivedAt < 0) {
            throw InboxMessageException(InboxMessageReason.UNREADABLE)
        }
        return InboxMessage(taskId, str("senderDeviceId"), body, receivedAt.toLong())
    }

    companion object {
        const val LABEL_PREFIX = "relayium/inbox/message"

        private const val SUFFIX = ".json"
        private const val VERSION = 1

        /** The protocol bounds a message at 64 KiB; the record adds only its
         *  small envelope, and JSON escaping cannot inflate it far. */
        private const val MAX_RECORD_BYTES = 1L * 1024 * 1024

        /**
         * What may be stored as a message.
         *
         * Bounded by the protocol's own text limits, and non-empty. Nothing is
         * repaired: a body that is not exactly what the sender wrote is not the
         * message, and showing a mended one would put words on screen that
         * nobody sent.
         */
        fun isAcceptable(text: String): Boolean {
            val bytes = text.toByteArray(Charsets.UTF_8).size.toLong()
            return bytes >= InboxManifest.MIN_TEXT_BYTES && bytes <= InboxManifest.MAX_TEXT_BYTES
        }
    }
}
